backend/src/main/java/gov/cdc/usds/simplereport/config/TenantDataAccessFilter.java

```diff
@@ -0,0 +1,70 @@
+package gov.cdc.usds.simplereport.config;
+
+import gov.cdc.usds.simplereport.api.CurrentTenantDataAccessContextHolder;
+import gov.cdc.usds.simplereport.api.model.errors.NonexistentUserException;
+import gov.cdc.usds.simplereport.config.authorization.TenantDataAuthenticationProvider;
+import gov.cdc.usds.simplereport.service.ApiUserService;
+import gov.cdc.usds.simplereport.service.AuthorizationService;
+import java.io.IOException;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.stereotype.Component;
+
+@Component
+public class TenantDataAccessFilter implements Filter {
+
+  @Autowired AuthorizationService _authService;
+  @Autowired ApiUserService _apiUserService;
+  @Autowired CurrentTenantDataAccessContextHolder _currentTenantDataAccessContextHolder;
+  @Autowired TenantDataAuthenticationProvider _tenantDataAuthProvider;
+
+  @Override
+  public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
+      throws IOException, ServletException {
+
+    Authentication currentAuth = SecurityContextHolder.getContext().getAuthentication();
+
+    try {
+      if (currentAuth != null && _authService.isSiteAdmin()) {
+        Optional<Set<String>> permissions =
+            _apiUserService.getTenantDataAccessAuthorityNamesForCurrentUser();
+
+        if (permissions.isPresent()) {
+          Set<GrantedAuthority> grantedAuthorities =
+              permissions.get().stream()
+                  .map(SimpleGrantedAuthority::new)
+                  .collect(Collectors.toSet());
+
+          String username =
+              _apiUserService.getCurrentApiUserInContainedTransaction().getLoginEmail();
+          _currentTenantDataAccessContextHolder.setTenantDataAccessAuthorities(
+              username, permissions.get());
+
+          // overwrite authentication with a new token so that subsequent request processing
+          // will see the authorities asserted by the active tenant data access rather than
+          // the token passed with the request
+          Authentication masqAuth =
+              _tenantDataAuthProvider.generateToken(username, currentAuth, grantedAuthorities);
```

Comment:
**question:** If we're overwriting the authentication token, why do we need the request-scoped context holder? I'm sure that I'm just misunderstanding the order of operations somewhere, but this implementation has me confused ☹️

Reply:
Covered more thoroughly in our discussion, but just so there's a little bit of record of it, the reason is because it is used to take precedence over Okta group membership when a site admin has active tenant data access claims.